tsdate/core.py

```diff
@@ -858,6 +931,129 @@ def outside_maximization(self, *, eps, progress=None):
         return self.lik.timepoints[np.array(maximized_node_times).astype("int")]
 
 
+class ExpectationPropagation(InOutAlgorithms):
+    """
+    Expectation propagation, where the edge factors are approximated
+    by the product of two gamma distributions
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        assert self.priors.probability_space == base.PAR
+        assert self.lik.probability_space == base.PAR
+        assert self.lik.grid_size == 2
+        assert self.priors.timepoints.size == 2
+
+        # Messages passed from factors in the direction of roots
+        self.parent_message = np.tile(
+            self.lik.identity_constant,
+            (self.ts.num_edges, 1),
+        )
+
+        # Messages passed from factors in the direction of leaves
+        self.child_message = np.tile(
+            self.lik.identity_constant,
+            (self.ts.num_edges, 1),
+        )
+
+        # Normalizing constants from each factor
+        self.factor_norm = np.full(self.ts.num_edges, 0.0)
+
+        # the approximate posterior marginals
+        self.posterior = self.priors.clone_with_new_data(
+            grid_data=self.priors.grid_data.copy(),
+            fixed_data=np.nan,
+        )
+
+        # factors for edges leading from fixed nodes are invariant
+        # and can be incorporated into the posterior beforehand
+        for edge in self.ts.edges():
+            if edge.child in self.fixednodes:
+                self.parent_message[edge.id] = self.lik.to_gamma(edge)
+                self.posterior[edge.parent] = self.lik.combine(
+                    self.posterior[edge.parent], self.parent_message[edge.id]
+                )
+                # self.factor_norm[edge.id] += ... # TODO
+
+    def edges_by_parent_asc(self):
+        """
+        Edges in order of increasing age of parent
+        """
+        return self.ts.edges()
+
+    def edges_by_child_desc(self):
+        """
+        Edges in order of decreasing age of child
+        """
+        wtype = np.dtype(
+            [
+                ("child_age", self.ts.tables.nodes.time.dtype),
+                ("child_node", self.ts.tables.edges.child.dtype),
+            ]
+        )
+        w = np.empty(self.ts.num_edges, dtype=wtype)
+        w["child_age"] = self.ts.tables.nodes.time[self.ts.tables.edges.child]
+        w["child_node"] = self.ts.tables.edges.child
+        sorted_child_parent = (
+            self.ts.edge(i)
+            for i in reversed(np.argsort(w, order=("child_age", "child_node")))
+        )
+        return sorted_child_parent
+
+    def propagate(self, *, edges, progress=None):
+        """
+        Update approximating factor for each edge
+        """
+        if progress is None:
+            progress = self.progress
+        # TODO: this will still converge if parallelized (potentially slower)
```

Comment:
Why potentially slower? Presumably because you would have to wait to update everything? @benjeffery has been doing some work over in tsinfer about parallelising work over independent edges, but my guess would be that in the case of tsdate, it's so fast as to not be worth the hassle working out the parallelisation scheme

Reply:
I just mean the rate of convergence will be slower, because the way in which edges are currently traversed (serially by time) is optimal. When parallelized, it'll break this order.

The algorithm overall might be much faster when parallelized, it just might take more iterations.